### PR TITLE
Performance and bug fixing :)

### PR DESCRIPTION
**Pull request by filbo**
I went looking for what was wrong with wsj.com, found so many issues,
this happened.  wsj.com works for me now on both Opera & Firefox.

- manifest-ff.json: add common.js to content script list
- contentScript, common.js: move matchDomain() from content to common
- common.js: make extension_api & defaultSites 'var' to fix redefinition errors on FF
- common.js: matchDomain() take an optional hostname (vs location.hostname)
- common.js: add urlHost() to extract hostname from URL
- common.js: add matchUrlDomain() to work on URLs (not appropriate on all calls)
- background.js: use matchUrlDomain() where appropriate
- background.js: string.includes() in place of indexOf() where appropriate
- background.js: check expensive isSiteEnabled() after cheap string compares
- background.js: check isSiteEnabled() *before* complex stuff that will abort anyway
- background.js: parse through entire requestHeaders[] fewer times
- background.js: no need to call handlerBehaviorChanged()
- background.js: inject common.js before contentScript.js, which now needs it

### CODE REVIEW

**Comment by filbo** [comment]
This needs & deserves thorough review, not quick commit...

Have tested on Firefox and Opera (Chrome-based), the following sites, with varied results (next comment).

**Comment by filbo** [comment]
I don't know which of these worked before (everything is dynamic since the sites themselves are constantly changing).

- bostonglobe.com -- problematic under Opera, works under FF
- cooking.nytimes.com -- couldn't get it to paywall me in any configuration.  Pictures always worked.  A shade, 'Oops! There was an issue submitting your request. Please try again.', comes from the top for a few seconds, only when BPW enabled; articles readable
- nytimes.com -- works.  A shade, 'Read more for free', pops from the bottom for a few seconds; articles readable
- medium.com -- works under FF; had trouble under Opera, articles led to 'error 500' page; however, enabling and disabling BPW cleared a cookie (or something), allowing me to read more articles
- towardsdatascience.com (uses medium) -- couldn't get it to paywall me
- theglobeandmail.com -- works
- wsj.com -- works
- sfchronicle.com -- works in conjunction with uBlock Origin, problematic without; didn't convince myself entirely that BPW was doing anything (vs. uBO carrying the entire load)
- techinasia.com -- works
- washingtonpost.com -- works.  A dismissable empty shade overlays part of the top; articles readable
- wired.com -- like sfchron: requires uBO, unclear whether uBO-alone was sufficient

Will retest problem areas under current checked-in version, but not for a day or two.

**Comment by pzamanis** [comment]
is not working in medium

**Comment by mrjohnson-tree** [comment]
> This needs & deserves thorough review, not quick commit...

Why commit untested changes though?
Or do you need a testpanel?

**Comment by bahari15** [comment]
> * wsj.com -- works

Is the wsj one still working for others?
I've tried clearing cache and cookies, and all the troubleshooting tips, but there's always 'Internal Server Error'

**Comment by frantzkamus** [comment]
> I don't know which of these worked before (everything is dynamic since the sites themselves are constantly changing).

Why are you proposing this PR if it breaks many websites?

**Comment by mrjohnson-tree** [comment]
Sabotage ...

**Comment by frantzkamus** [comment]
No offence, but I find it weird that you have titled your PR 'bug fixing' while recognizing at the same time that it'll break many websites.

**Comment by filbo** [comment]
@frantzkamus:

The only sites that _didn't work_ for me were bostonglobe and medium, only under Opera (therefore presumably under any Chromium-browser).  medium, while not directly working, did allow me to read as many articles as I wanted (the purpose of the extension), by enabling and disabling it periodically to clear cookies.  Furthermore, as I said, I don't know what their statuses were under the prior code; that is for future testing.  'will break many websites' is a mischaracterization.

@mrjohnson-tree:

> Why commit untested changes though?

To get eyes on them?  To get further testing?  These changes are committed into a branch of my tree, not into @iamadamdev's mainline.  As I said, they _should not_ be committed into mainline until further examined.

> Or do you need a testpanel?

I am not sure whether you mean a test suite (some sort of automated or manual set of testing procedures); or a group of people to do testing.  Both would be helpful.

@pzamanis:

> is not working in medium

and this is a regression for you, with my code, vs the current top of @iamadamdev's tree?

@bahari15:

> Is the wsj one still working for others?
> always 'Internal Server Error'

Is that on the main page wsj.com; articles; or what?  Please state browser info; exact problem URL (one is enough); entire error message.

**Comment by mrjohnson-tree** [comment]
Your commits also show up in this PR.
If merged we would have acted as crash test dummies ...
You're not concerned about possible legal issues while using your own name here?

**Comment by filbo** [comment]
They are commits to my branch of my fork!  I don't see what your concern is with that at all.  Of course if Adam merges the PR into his upstream tree, then it is a concern.

I'm presenting the code here **so it can be tested**, as well as reviewed by eye or other means.  It is a PR because I want to submit the eventual product of all such review into Adam's tree.  Without submitting it as a PR, I know of no good way to ensure attention (in terms of both eyes & testers) to the code.

Now, tonight, I sit down to retest some problem sites, compare results between current upstream top-of-tree and my version, etc.  Note, the sites I concentrated my testing on were precisely those which seemed most likely to be affected, as they had their own specialized bits of code.  Since I haven't yet done the regression testing, I don't know if the 2-sites-on-1-browser which didn't work are worse than before (vs. the sites having changed, breaking their site-specific bypass algorithms).

**Comment by filbo** [comment]
> You're not concerned about possible legal issues while using your own name here?

Maybe, slightly, vaguely?  I actually feel like getting sued over a project like this would be one of those 'all publicity is good publicity' situations.  The party laying the suit would end up looking like a bunch of schmucks and I'd get a nice job out of it, working on something fun :)

**Comment by filbo** [comment]
> No offence, but I find it weird that you have titled your PR 'bug fixing' while recognizing at the same time that it'll break many websites.

Re-replying to this: it is my experience that all software maintenance work involves the addition of new bugs.  The trick is to make sure you are only adding tiny bugs while fixing big ones.  Over time, the fractal bug-surface of the project is whittled down into a finer and finer texture of tiny bugs with (hopefully) no big ones sticking way out where they could hurt someone.

By raising the PR **WITH** notes about breakage in it, I was clearly signaling that this is not to be committed as-is, but to be studied, improved, until it doesn't have those larger more visible defects.  As I stated in the initial comments.

This does mean I put some expectations on the main owner of the project: that he pays attention to the shape of submissions, doesn't commit them willy-nilly.  My first submission seemed to follow a reasonable trajectory, with days of conversation and several fixes along the way before he committed it; expectations met.

I would rather this be collaborative, not antagonistic.  If you see new problems with a site, perhaps you could look at _why_ it has problems and propose how to fix them, rather than sniping about people's motivations.

**Comment by mrjohnson-tree** [comment]
> They are commits to my branch of my fork! I don't see what your concern is with that at all. Of course if Adam merges the PR into his upstream tree, then it is a concern.

Yes, like the previous commited and merged errors in contentscript ...
I don't expect you to fix previously not working sites, but at least testing your own commits would be 'nice'.

> I would rather this be collaborative, not antagonistic. If you see new problems with a site, perhaps you could look at why it has problems and propose how to fix them, rather than sniping about people's motivations.

We're looking forward for you to fix some sites.
So far mainly "cleaner code' and 'performance improvements' (tiny compared to loading time of average site though).

> I would rather this be collaborative, not antagonistic. If you see new problems with a site, perhaps you could look at why it has problems and propose how to fix them, rather than sniping about people's motivations.

A bit difficult without an issues section ...

**Comment by filbo** [comment]
>> If you see new problems with a site, perhaps you could look at why it has problems and propose how to fix them, rather than sniping about people's motivations.

> A bit difficult without an issues section ...

Well, I meant problems introduced in a proposed PR.  The 'issues section' for that is the discussion area under the PR: here.

Of course if you see problems with other sites, unrelated to a pending PR, you can also raise 'issues' by submitting your own PR which purports to fix the problem.  I agree the Issues section should be enabled, but it is not entirely debilitating.

File: common.js [hunk 1]
```diff
@@ -1,4 +1,4 @@
-const extension_api =
+var extension_api =
```
Comment (iamadamdev) [inline common.js:1]: Where are you seeing errors for this in FF?
In about:debugging after loading the extension, I clicked Inspect, and in the console I didn't see any errors.

File: common.js [hunk 2]
```diff
@@ -135,3 +135,24 @@ const defaultSites = {
...
+    if (typeof domains === 'string')
+        domains = [domains];
+    domains.some(domain => (hostname === domain || hostname.endsWith('.' + domain)) && (matched_domain = domain));
+    return matched_domain;
+}
+
+function urlHost(url) {
+    try {
+        return new URL(url).hostname;
+    } catch(e) { e; }
```
Comment (iamadamdev) [inline common.js:152]: No-op or log to console?

**Comment by wick3d212** [comment]
not working on medium. logged out and cant log in again. any fix?

**Comment by iamadamdev** [comment]
1. Rebase off latest master
2. If you want, another small optimization to make is in contentScript make all the if checks if-else checks

**Comment by iamadamdev** [comment]
@wick3d212 Which medium article isn't working for you? It's not supposed to keep you logged in to any site. You must uncheck the site if you want to use your own login.
Statista still works fine for me in Chrome. Which doesn't work for you?
Any other sites don't work for you? Please paste exact articles.
thanks

**Comment by wick3d212** [comment]
Well medium.com doesnt let me in. Cant see anything other then main page where tags and stuff is. And other sites i checked from posted list only few works others dont. Other than that i can see anything. Statista does it let us download xls? 

Medium.com and statista which i really like to work areñt working at all for me. I tried firefox and chrome both. Same thing. Doesnt work.

**Comment by iamadamdev** [comment]
@wick3d212 Try articles from https://topmediumstories.com/
Statista just shows a graph. I don't think downloading the xls ever worked even on master branch.

**Comment by wick3d212** [comment]
Yea it shows top stories. But i was thinking new stories everyday. Can only happen if that extension works on that site. :( could be something creating that problem? Some other extension maybe?

**Comment by iamadamdev** [comment]
I believe medium changed their landing page. Open incognito mode without any extensions and you can see that.

**Comment by mrjohnson-tree** [comment]
> I believe medium changed their landing page. Open incognito mode without any extensions and you can see that.

Medium or Topmediumstories?
Medium has had this page for months.

**Comment by mrjohnson-tree** [comment]
Not working:
https://www.wsj.com/articles/state-coronavirus-shutdowns-have-taken-29-of-u-s-economy-offline-11586079001?mod=mhp

File: common.js (hunk 1 above)
Comment (filbo) [reply common.js:1]: Did you check the JS console of the visited pages (not the extension)?  I forget at this point, but I think it was coming out of the content script.  I'll comment further if I reproduce it.

File: common.js (hunk 2 above)
Comment (filbo) [reply common.js:152]: No-op as in `catch(e) { }` ?  JS balks at `catch() { }` (arg is required).  JS linters I've used, although not on this, get mad about the 'unused variable' in `catch(e) { }`.  So I refer to it.

But yeah, log is better: `catch(e) { console.log('urlHost fails:', url, e); }` or something like that.

**Comment by filbo** [comment]
> Although I don't have extract repro steps yet, I noticed I did get this error after browsing a few sites and went back to the chrome://extensions/ tab
> `Unchecked runtime.lastError: Cannot access contents of the page. Extension manifest must request permission to access the respective host.`

Hmmm, tell me if you learn any more about that.

> 1. Rebase off latest master

Will do.  Was the removal of `'ed.nl': ['temptationTrackingId'],` from remove_cookies_select_drop{} intentional?

> 2. If you want, another small optimization to make is in contentScript make all the if checks if-else checks

I have something else in mind for performance along that axis, so will skip that particular change.

**Comment by johnycomp11** [comment]
I added the file to Chrome extension but it is not working for Seeking Alpha.

**Comment by mrjohnson-tree** [comment]
> Was the removal of 'ed.nl': ['temptationTrackingId'], from remove_cookies_select_drop{} intentional?

Another copied and untested commit ...?
Is it just me or does it looks like more bypasses are failing?

**Comment by berryman13** [comment]
This version doesn't work on WSJ, same error as the original.

**Comment by iamadamdev** [comment]
WSJ works fine for me on this PR and master. Tested on https://www.wsj.com/articles/state-coronavirus-shutdowns-have-taken-29-of-u-s-economy-offline-11586079001?mod=mhp

**Comment by iamadamdev** [comment]
> Will do. Was the removal of `'ed.nl': ['temptationTrackingId'],` from remove_cookies_select_drop{} intentional?
Yes, it's not needed on ed.nl anymore.

**Comment by mrjohnson-tree** [comment]
Doesn't sound right when it's still needed on ad.nl (same cookiename).

WSJ randomly fails, but link above is working for me now too.
Fails: https://www.wsj.com/articles/with-online-package-delivery-alerts-later-in-the-process-is-better-11585352291

**Comment by iamadamdev** [comment]
> Doesn't sound right when it's still needed on ad.nl (same cookiename).
> 
> WSJ randomly fails, but link above is working for me now too.
> Fails: https://www.wsj.com/articles/with-online-package-delivery-alerts-later-in-the-process-is-better-11585352291

Which article on ed.nl doesn't work for you? And does adding that line back fix it?

**Comment by mrjohnson-tree** [comment]
Why did you erase the line when you clearly didn't test it?

**Comment by iamadamdev** [comment]
Which article on ed.nl doesn't work for you?
I tested it. They all worked for me.

**Comment by mrjohnson-tree** [comment]
Sure, only full cookiebanner on all pages.
Still why the change?

**Comment by iamadamdev** [comment]
Works fine for me.

**Comment by mrjohnson-tree** [comment]
With another extension (blocking cookiebanners) it does.
Still strange you changed it during a commit for another site, but you don't want to go into that ...

**Comment by filbo** [comment]
@mrjohnson-tree your sniping is tiresome; I better understand @iamadamdev's past moves.

> Fails: https://www.wsj.com/articles/with-online-package-delivery-alerts-later-in-the-process-is-better-11585352291

I can reproduce that.  Adam:

> WSJ works fine for me on this PR and master. Tested on https://www.wsj.com/articles/state-coronavirus-shutdowns-have-taken-29-of-u-s-economy-offline-11586079001?mod=mhp

-- that might not be a good test; many sites are putting their covid-19 related materials outside their paywalls.  I don't know whether WSJ is, just in general it's a good idea to look for unrelated articles (if you can find any).

I tried a bunch of articles on WSJ and they all work -- covid-19 or completely unrelated -- except the one pointed out by our buddy.  So yeah, something is wrong; no, it's not immediately obvious.

**Comment by mrjohnson-tree** [comment]
No 'sniping' needed if Adam could just say he copied a fix by @magnolia1234 again: https://github.com/magnolia1234/bypass-paywalls-chrome-clean/commit/f7506805b3f38d63b389809e8d148c990a39a6b7

About WSJ: random paywall on some articles with a video at the top (timing issue?).

**Comment by filbo** [comment]
About WSJ: testing the same two articles, 

https://www.wsj.com/articles/with-online-package-delivery-alerts-later-in-the-process-is-better-11585352291

-- fails on BOTH this PR *and* bypass-firewalls-clean.

https://www.wsj.com/articles/state-coronavirus-shutdowns-have-taken-29-of-u-s-economy-offline-11586079001?mod=mhp

-- works with both.

And, both articles fail with the current git version of bypass-paywalls-chrome@iamadamdev.  For me, on Opera.  It's clear that these things are not consistent from browser to browser, possibly client IP to IP, and may also have temporal or other components.

To be clear:

11585352291 'package delivery':  @iamadamdev fail, @magnolia1234 fail, this PR fail
11586079001 'economy offline':  @iamadamdev fail, @magnolia1234 succeed, this PR succeed

**Comment by mrjohnson-tree** [comment]
WSJ still has Cxense paywall (visible for recommendations), but blocked by uBO.
Maybe also active on other articles.

**Comment by teajoe** [comment]
also please add kansas city star newspaper into extension if possible,
thanks
Respectful
Ming Zhong

**Comment by iamadamdev** [comment]
@teajoe Added in latest master

**Comment by bradwolff18** [comment]
Cant submit new website request, but can you add SportsBusiness Journal please?

**Comment by hikeith** [comment]
Hey bro, Both ET Prime and WSJ is not working, Tried several times but nothing is working, can you just look into it.

**Comment by pcaversaccio** [comment]
Please add "Der Spiegel" (https://www.spiegel.de/) & "Frankfurter Allgemeine" (https://www.faz.net/aktuell/) into the extension if possible. Those are well-known German Newspapers and are frequently referred in FT/WSJ Articles.
Thanks a lot.
Best,
Pascal

**Comment by kimkha** [comment]
@pcaversaccio I think https://www.faz.net/aktuell/ has no paywalls. It should work without this extension. You may need to disable adblock (if you have).

btw, @iamadamdev Why do you turn off the Issues tab? People keep messing this PR with the website request...

**Comment by pcaversaccio** [comment]
@kimkha that's just the general URL. Check out for instance: https://www.faz.net/aktuell/wirtschaft/die-corona-exit-strategie-des-sachverstaendigenrats-16721404.html - FAZ has a soft paywall.

**Comment by iamadamdev** [comment]
@kimkha People were abusing it and spamming links here so I had to turn it off.

**Comment by iamadamdev** [comment]
@bradwolff18 sportsbusinessdaily.com has a hard paywall.
@hikeith ET prime is not a supported site. It was removed some time ago. Some articles on WSJ are not working, that is a known issue.
@pcaversaccio Both those sites have a hard paywall. (You can test this by disabling javascript for example)

File: common.js (hunk 1 above)
Comment (iamadamdev) [reply common.js:1]: Yes, still not seeing any errors related to const vs var.

**Comment by iamadamdev** [comment]
@filbo Do you mind separating this out into separate PRs?
It will be easier to test and revert if necessary.
I am a bit hesitant to merge this since I am seeing an error on Chrome that I don't see on latest master.
Also I believe the main purpose of this PR was to fix WSJ? Does this PR fix articles that don't work on latest master?
There are known issues with WSJ lately in that some articles are not loading, no fix yet.

File: common.js (hunk 1 above)
Comment (filbo) [reply common.js:1]: @iamadamdev -- would you please contact me via email (my addresses are easily found; use gmail).  Things to discuss without so much noise.

**Comment by filbo** [comment]
>  seeing an error on Chrome that I don't see on latest master

What is it?  Of course don't merge until issues are resolved.

**Comment by pipiscrew** [comment]
I dont know where to post it, so merging on this, new kid @ paywalls :
https://beta.trimread.com/

**Comment by kimkha** [comment]
@pipiscrew are you trying to advertise that website? You better post via reddit, twitter or website instead, and please don't post here.

**Comment by SALLU99Ch** [comment]
Hi guys, I'm new on this community, and I don't know very well how to send a request to include a new site on the extension. Can you help me?
The site is "Il Sole 24 Ore". It's very famous in Italy.
Thank you.

**Comment by kimkha** [comment]
@SALLU99Ch I don't think this extension could help. The site https://24plus.ilsole24ore.com has hard paywall, while https://www.ilsole24ore.com has no paywall. This extension just can't bypass the hard paywall.

**Comment by wick3d212** [comment]
Appreciate dev for a bit guys. Hez already done a marvelous job. And what we all r doing just pushing him to add this and that. Comeon man...

**Comment by SALLU99Ch** [comment]
> @SALLU99Ch I don't think this extension could help. The site https://24plus.ilsole24ore.com has hard paywall, while https://www.ilsole24ore.com has no paywall. This extension just can't bypass the hard paywall.

I understand, but how can I know if a paywall is a hard paywall or not? Only with the method written in the README? I ask this because the site "La Repubblica" has a very similar system to "Il Sole 24 Ore" and in the first case the paywall can be avoided without a problem.
In any case, thank you for answering!

**Comment by kimkha** [comment]
The reason why the site `La Repubblica` could be bypassed is that there is some work around like this:

https://github.com/iamadamdev/bypass-paywalls-chrome/blob/master/contentScript.js#L10-L15

and this (which is blocking some requests):

https://github.com/iamadamdev/bypass-paywalls-chrome/blob/master/background.js#L175

I said it has a hard paywall, because I did research and `Il Sole 24 Ore` doesn't look like the same as `La Repubblica` in term of code. It's sending a graphql request to its server, the server will response the full content or not based on your current login. But I may be wrong.

Can you find some work around to get rid of its paywall?

If you can't, the extension can't, I'd simply say that.

**Comment by afmiquel** [comment]
Hi not sure this the right place to say this, but the extension doesn't seem to work with Le Monde paywall

**Comment by galafrin** [comment]
Congratulation for the great work! 

I noticed that Lemonde.fr can't pass thru its paywall since a couple of weeks either from FF or Chrome, and following the guidelines tricks brings no fruition eitjer yet I still can pass the paywalls of WSJ and NYT . So I deduce that something must have changed in the way Lemonde.fr processes general public's requests. 

I wonder if you are using any step-by-step HtML-script debugger to sort issues like this out?

**Comment by filbo** [comment]
> @filbo Do you mind separating this out into separate PRs?

@iamadamdev did you see my request to contact me via email?

**Comment by iamadamdev** [comment]
I emailed you a few days ago, I guess you didn’t get it? *_*@ar****.com

**Comment by filbo** [comment]
Ah, I asked you to use gmail.  And I do check that other account, and -- now I see it, must have missed it earlier in the welter of copy-to-email comments from this PR and others.  Will reply there.

**Comment by farmermansoft** [comment]
Great tool - thanks!  I am having problems with NYT - with adblocker enabled and disabled.  Chrome browser, Windows 10.   The website keeps refreshing and ads and sections move around. I end up copy and paste content into a Word document to be able to read it.

**Comment by iamadamdev** [comment]
@filbo Lets discuss offline, I sent you an email again at your armory address. I also tried sending to a gmail address but it bounced back.

**Pull request closed without merging**